### PR TITLE
Add raw body bytes capability

### PR DESCRIPTION
**Pull request by maiqbal11**
Fixes https://github.com/Azure/azure-functions-host/issues/4556

### CODE REVIEW

File: src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs [hunk 1]
```diff
@@ -196,6 +214,21 @@ public static TypedData ToRpc(this object value, ILogger logger)
             return typedData;
         }
 
+        private static bool RawBodyBytesRequested(Capabilities capabilities)
+        {
+            return string.Equals(capabilities.GetCapabilityState(LanguageWorkerConstants.RawHttpBodyBytes),
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs]: Only check if the state is not null. A language worker would send capability
```
{"RawHttpBodyBytes", "state can be json in future if needed."}
```
An older version of  python worker would simple not send this capability on init

File: src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs [hunk 2]
```diff
@@ -196,6 +214,21 @@ public static TypedData ToRpc(this object value, ILogger logger)
             return typedData;
         }
 
+        private static bool RawBodyBytesRequested(Capabilities capabilities)
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs]: Rename: IsRawBodyBytesRequested

File: src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs [hunk 3]
```diff
@@ -141,7 +141,9 @@ public static TypedData ToRpc(this object value, ILogger logger)
                 if (request.Body != null && request.ContentLength > 0)
                 {
                     object body = null;
+                    object rawBodyBytes = null;
                     string rawBody = null;
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs]: Rename rawBody  to rawBodyString

File: src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs [hunk 4]
```diff
@@ -177,8 +183,20 @@ public static TypedData ToRpc(this object value, ILogger logger)
...
+                        http.RawBody = rawBodyBytes.ToRpc(logger, capabilities);
+                    }
+                    else
+                    {
+                        http.RawBody = rawBody.ToRpc(logger, capabilities);
+                    }
+
+                    if (rawBodyBytes == null && RawBodyBytesRequested(capabilities))
+                    {
+                        http.RawBody = bytes.ToRpc(logger, capabilities);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs]: If..else blocks should be enough. You do not need `rawBodyBytes`
```
   if (RawBodyBytesRequested(capabilities))
                    {
                        http.RawBody = bytes.ToRpc(logger, capabilities);
                    }
                    else
                    {
                        http.RawBody = rawBody.ToRpc(logger, capabilities);
                    }
```

File: test/WebJobs.Script.Tests/Rpc/RpcMessageConversionExtensionsTests.cs [hunk 5]
```diff
@@ -18,17 +21,20 @@ namespace Microsoft.Azure.WebJobs.Script.Tests.Rpc
...
         {
             var logger = MockNullLoggerFactory.CreateLogger();
+            var capabilities = new Capabilities(logger);
 
             var headers = new HeaderDictionary();
             headers.Add("content-type", expectedContentType);
             HttpRequest request = HttpTestHelpers.CreateHttpRequest("GET", "http://localhost/api/httptrigger-scenarios", headers, body);
 
-            var rpcRequestObject = request.ToRpc(logger);
+            var rpcRequestObject = request.ToRpc(logger, capabilities);
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Rpc/RpcMessageConversionExtensionsTests.cs]: Verify HttpRawBody is string

File: test/WebJobs.Script.Tests/Rpc/RpcMessageConversionExtensionsTests.cs [hunk 6]
```diff
@@ -247,5 +255,49 @@ internal static ClaimsIdentity MockEasyAuth(string provider, string name, string
...
+            HttpRequest request = HttpTestHelpers.CreateHttpRequest("GET", "http://localhost/api/httptrigger-scenarios", headers, imageBytes);
+
+            var rpcRequestObject = request.ToRpc(logger, capabilities);
+            if (rawBytesEnabled)
+            {
+                Assert.Equal(imageLength, rpcRequestObject.Http.RawBody.Bytes.ToByteArray().Length);
+            }
+            else
+            {
+                Assert.NotEqual(imageLength, rpcRequestObject.Http.RawBody.Bytes.ToByteArray().Length);
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Rpc/RpcMessageConversionExtensionsTests.cs]: Isn't this a string?

Comment (maiqbal11) [reply test/WebJobs.Script.Tests/Rpc/RpcMessageConversionExtensionsTests.cs]: Yes, you're right. Added a test to check that the string length is equal to what we expect from the ToRpc function.

File: test/WebJobs.Script.Tests/Rpc/RpcMessageConversionExtensionsTests.cs (hunk 5 above)
Comment (maiqbal11) [reply test/WebJobs.Script.Tests/Rpc/RpcMessageConversionExtensionsTests.cs]: Should be part of checking that the length of the string is what we expect.

File: src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs [hunk 7]
```diff
@@ -196,6 +209,20 @@ public static TypedData ToRpc(this object value, ILogger logger)
             return typedData;
         }
 
+        private static bool IsRawBodyBytesRequested(Capabilities capabilities)
+        {
+            return !string.Equals(capabilities.GetCapabilityState(LanguageWorkerConstants.RawHttpBodyBytes), null);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs]: change to `return capabilities.GetCapabilityState(LanguageWorkerConstants.RawHttpBodyBytes) != null;`

File: test/WebJobs.Script.Tests/Rpc/RpcMessageConversionExtensionsTests.cs [hunk 8]
```diff
@@ -247,5 +255,53 @@ internal static ClaimsIdentity MockEasyAuth(string provider, string name, string
...
+
+            long imageBytesLength = imageBytes.Length;
+            long imageStringLength = imageString.Length;
+
+            var headers = new HeaderDictionary();
+            headers.Add("content-type", contentType);
+            HttpRequest request = HttpTestHelpers.CreateHttpRequest("GET", "http://localhost/api/httptrigger-scenarios", headers, imageBytes);
+
+            var rpcRequestObject = request.ToRpc(logger, capabilities);
+            if (!string.Equals(rawBytesEnabled, null))
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Rpc/RpcMessageConversionExtensionsTests.cs]: Update 
```
   if (string.IsNullOrEmpty(rawBytesEnabled))
            {
                Assert.Empty(rpcRequestObject.Http.RawBody.Bytes);
                Assert.Equal(imageStringLength, rpcRequestObject.Http.RawBody.String.Length);
            }
            else
            {
                Assert.Empty(rpcRequestObject.Http.RawBody.String);
                Assert.Equal(imageBytesLength, rpcRequestObject.Http.RawBody.Bytes.ToByteArray().Length);
            }
```

File: src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs [hunk 9]
```diff
@@ -141,44 +141,57 @@ public static TypedData ToRpc(this object value, ILogger logger)
                 if (request.Body != null && request.ContentLength > 0)
                 {
                     object body = null;
-                    string rawBody = null;
+                    object rawBodyBytes = null;
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs]: why do you need rawBodyBytes?

Comment (maiqbal11) [reply src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs]: Shouldn't be needed. I think it just got left over while I was iterating on the code logic.

**Pull request merged**